Automattic/wp-calypso
# Checkout: Refactor WeChatPay to move modal to processor

## Proposed Changes

The WeChatPay payment method has an unusual flow. When the user clicks to pay, we must display a custom modal dialog with a QR code. When the user confirms the purchase via an external site, a webhook will mark the order as successful or failed. While the QR code is visible, checkout polls for that webhook to determine the status of the order before sending the user to the post-checkout page or back to checkout on a failure.

However, there are a few problems with this approach. In order to perform the above dance, the payment processor function must notify the payment method code to display the modal. The modal then must manually redirect after checking the status. This mixes the intended purposes of payment method and payment processor function; the payment method is only supposed to set up the purchase flow for the user and the payment processor function is supposed to do all the heavy lifting. Because of this, the checkout payment system has to have a "manual" processor response and the payment method has to perform its own redirects.

In this PR, we refactor WeChatPay to instead move the modal entirely over to the payment processor function. It then uses the standard payment processor responses to handle success and failure.

<img width="556" alt="Screenshot 2024-01-23 at 7 53 25 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/145c40f0-a6aa-4bd7-a88a-4b957efcb04b">

> [!NOTE]
> This uses a `dialog` HTML element to display the QR code in order to prevent fighting the `CheckoutSlowProcessingNotice` overlay for z-index height. As of this writing, `dialog` is well-supported in the browsers calypso supports, but there is still a possibility that it may not work for someone. If we ever want to go back to using a custom dialog, there's a commit in this PR that can be undone with a few adjustments, but we'll need to figure out a way to either make this dialog have a z-index above `CheckoutSlowProcessingNotice` (challenging since they are in different stacking contexts) or prevent `CheckoutSlowProcessingNotice` from being rendered while we are waiting on the dialog (if we could find a way to know when the actual call to the transactions endpoint completes).

## Testing Instructions

- Set your currency to USD.
- Apply D45650-code or a similar version to force WeChatPay availability and sandbox the API.
- Add a product to your cart and visit checkout.
- Select "WeChat Pay" as the payment method.
- Fill out the "Name" field and submit the payment.
- In **a new window or tab** open the link below the QR code that reads "To open and pay with the wechat pay app directly, click here".
- In the new window or tab, click to fail the payment. Then **return to the original window**.
- Verify that the modal disappears and that an error message is shown.
- Repeat the payment steps but this time click to confirm the payment.
- Verify that you are redirected to the post-checkout flow.
